### ISSUE TITLE
[Discussion] More statements as expressions.

### ISSUE THREAD

**Issue by orthoxerox** [dotnet/roslyn#5882]
Since this was briefly mentioned in #5154 I've decided to create a separate issue for this discussion.

**1. Blocks inside method bodies are expressions and return the value of their last expression.**

```c#
let x = {
	let a = 1;
	let b = 2;
	a + b;
}
```

**2. You cannot end a block inside a method body with a declaration or a statement that isn't guaranteed to not return if its return value is used.**

```c#
let x = {
	let a = 1;
	let b = 2;
	a + b;
	int y; //I won't compile!
}
```

**3. `if` is now an expression and accepts expressions as its two branches and returns the value of the one that is executed. When the return value of `if` is used its `else` branch becomes mandatory.**

```c#
let x =
	if (list != null && list.Count > 0) {
		list[list.Count - 1];
	} else {
		getSomeDefaultValue();
	}

let y =
	if (list != null && list.Count > 0) {
		list[0];
	} //I won't compile!
```

**4. `for`, `foreach`, `while` and `do ... while` are not expressions (because their body can be executed zero times).**

**5. `try` and its derivatives are now expressions and return the value of the `try` block when no exception is thrown and the value of a `catch` block when an exception is caught by one. The value of the `finally` block is discarded.**

```c#
let result =
	try {
		GetResult();
	}
	catch (Exception ex) {
		logger.Log(ex);
		GetSomeDefaultResult();
	}
	finally {
		Cleanup();
	}
```

```c#
return lock(lockField)
	this.blahField;
```

**6. `switch` is not an expression (because there's `match`.)**

**7. You cannot yield a value of a non-final expression in a block.**

**Comment by alrz** [comment]
basically, #5402

**Comment by orthoxerox** [comment]
Duplicate of #5402
